### ISSUE TITLE
[PPconverter] Mapping rules for casting

### ISSUE THREAD

**Issue by pasqLisena** [DOREMUS-ANR/marc2rdf#22]
In the mapping rules for the Philarmonie, all properties for casting (except a textual `P3 has note`) are marked with 

> (non exploité : pas d'équivalent à la source)

But we are wasting a lot of details that PP has in their data.

### Examples

Here some examples coming from Beethoven and Dutilleux's files.

```
Piano et orchestre
2 violons, alto, violoncelle
2 petites flûtes (prenant aussi la grande flûte), 2 grandes flûtes, 3 hautbois, cor anglais, petite clarinette (en mib), 2 clarinettes (en sib), clarinette basse (en sib), 3 bassons, contrebasson, 4 cors, 4 trompettes, 3 trombones, tuba, 4 timbales, percussion (2 temple-block, caisse claire, 3 toms -aigu, médium, grave- grosse caisse à pied, petite cymbale suspendue, cymbale chinoise, tam-tam médium, tam-tam grave, cymbales frappées, triangle, cow-bell), xylophone, glockenspiel, celesta, harpe, quinette à cordes. Editeur : Heugel
Violoncelle
2 ensembles : 1 groupe de 12 instruments solistes (1 hautbois, 1 clarinette, 1 basson, 1 trompette, 1 trombone, 1 clavecin, 1 celesta, timbales, 2 violons, 1 alto et 1 violoncelle); 1 grand orchestre (3 flûtes, 2 hautbois et 1 cor anglais, 2 clarinettes et 1 clarinette basse, 2 bassons et 1 contrebasson, 2 cors, 2 trompettes, 2 trombones, 1 tuba, percussions, harpe et quintette à cordes)
Violoncelle et orchestre
Violon, orchestre
Piano
24 cordes, cymbalum et percussions
Hautbois, clavecin, contrebasse et percussion
Piano
1 flûte piccolo, 2 flûtes, 2 hautbois, cor anglais, 2 clarinettes en si bémol, clarinette basse en si bémol, 2 bassons, 3 cors en fa, 3 trompettes en ut, 3 trombones, tuba, timbales, percussions (2 cymbales suspendues, tam-tam, 3 bongos, 3 tom-toms, caisse claire, grosse caisse), marimbaphone (aussi vibraphone), accordéon, célesta, harpe et cordes.
Violon et orchestre
```

### What we can do

- Split the string (commas `,` , the word `et` and brackets `(  )` )
- Clean data (`Editeur : Heugel` should not be there)
- Save the number of each mop
- Leave it as comment when we are not able to identify a pattern

### Critical points

1. The percussionist often plays different percussions.
2. We have never thought how to represent ensamble members: how to specify that is an `orchestra` and in the same time declare the instruments that it includes?
3. The case of `2 ensambles` is manifestly the most difficult.

Maybe this critical points are not related to the implementation, but to the modelling.

**Comment by pierrechoffe** [comment]
@marie-ototoi @pasqLisena sorry for this late answer... :) About critical points

>The percussionist often plays different percussions

Why is it a problem ? the casting detail describes the instruments, not the players. So if we have 3 instruments played by 1 person only, there will be 3 casting details.

>We have never thought how to represent ensemble members

The casting detail can represent a group (orchestra, etc.), e.g. if we have a concerto for violin and orchestra (without further details), we have 2 casting details : violin, orchestra.
If we know more about this group, as in the example given by Marie, then we can say that the casting detail [ensemble] p69_has_association_with 12 casting details representing the 12 soloists.
To be more precise, we should maybe create a subproperty of p69 with an adequate name.

>The case of 2 ensembles is probably the most difficult

Using the above model, we have no problem with any number of ensembles.

But yes, this leaves open the question of the number of players, which should be addressed.

**Comment by pasqLisena** [comment]
> > The percussionist often plays different percussions

> Why is it a problem ? the casting detail describes the instruments, not the players. So if we have 3 instruments played by 1 person only, there will be 3 casting details.

Sometimes we have the sheets separated for percussionists (i.e. _Percussion 1_ plays triangle and bass drum, _Percussion 2_ cymbals and tubolar bells, _Percussion 3_ timpani). It apply also to other instruments (i.e. often the 2nd Flute plays the Piccolo for few measures).

I don't know if we have this info in our data. But, in any case, this could not be an information to keep?

> we can say that the casting detail [ensemble] p69_has_association_with 12 casting details representing the 12 solists

Great! This is a solution!

> we should maybe create a subproperty of p69 with an adequate name

Yes, like `uxx_has_member`.

> > The case of 2 ensembles is probably the most difficult
> Using the above model, we have no problem with any number of ensembles.

I agree. However it is the most complex example we have from the parsing point of view.

> But yes, this leaves open the question of the number of players, which should be addressed.

I am not understanding what you are referring to.

**Comment by pierrechoffe** [comment]
@pasqLisena 

>I don't know if we have this info in our data. But, in any case, this could not be an information to keep?

Yes, this is very interesting but maybe we enter too much into details... We had this kind of questions in the beginning of the project, when we thought that the model could help the music library of Radio France for example, but RF said it would raise other problems, so we dropped the idea. As you know, a score may be divided between interpreters, but then you have other parameters that enter into account, sometimes not really musical ones...

>this leaves open the question of the number of players, which should be addressed

We had talked about the description of a "number of players" when this is explicitly mentioned by the composer, e.g. a piece "for 2 pianos or 4-hands piano". In this case we have an alternate casting : casting 1 is straightforward (mop=piano, quantity=2), casting 2 involves counting the number of performers (mop=piano, quantity=1, number_of_performers=2).
This means that unless explicitly mentioned, the number of players is 1.

If the implicit number of players is a problem, then we have to model it differently, not on M23_Casting but as another intention (E29), pointing to the related M23.

**Comment by marie-ototoi** [comment]
Nous avions estimé que ce champ n'était pas renseigné de façon assez structurée et systématique pour être exploitable... mais peut-être qu'en le comparant avec le référentiel des MOP on peut ne garder que ce qui matche ?

**Comment by rtroncy** [comment]
@pasqLisena Would you like to give it a try in the MARC2RDF implementation, and we will manually review what has been extracted for the corpora we have been given (or samples of it)?

**Comment by pasqLisena** [comment]
Done:
- [x] Not complex sets of instruments

examples:
```
2 violons, alto, violoncelle
Flûte, clarinette, vibraphone, piano, violon, violoncelle
ténor solo, baryton solo, choeur, 3 flutes dont 1 piccolo, 3 hautbois, 3 clarinettes dont 1 clarinette basse, 3 bassons dont 1 contrebasson, 4 cors, 2 trompettes, 3 trombones, tuba basse, percussion, harpe, cordes
Voix-dessus (2), (solo), voix-basse-taille (1), (solo), clavier-piano (1)
```

Todo:
- [ ] Castings with roles
```
Rôles : Nymphe de la Seine, La Gloire, Nymphe des Tuileries, Nymphe de la Marne, Alceste, Céphise, Thétis, Proserpine, Esprit repoussé, Diane (sopranos) ; Admète, Lychas, Apollon, Alecton (haute-contre) ; Cléante, Phérès (ténors) ; Alcide, Charon, Eole (barytons) ; Straton, Pluton (basse). Choeurs, orchestre symphonique.
```
- [ ] Complex castings and orchestras
```
Ouverture : les bois par 2, 2 cors, 2 trompettes et 1 tuba, timales et quintette à cordes. Musique de scène : pour deux sopranos, choeur féminin et orchestre ( les bois par 2, 2 cors, 3 trompettes, 3 trombones, 1 ophicléide ou tuba, timbales, petite batterie et quintette à cordes
```
- [ ] singularize terms for vocabulary matching
- [ ] manage the alternative castings (`1 ophicléide ou tuba`)


I add this list also on the top.
